Source/WebKit/Shared/JavaScriptEvaluationResult.mm

```diff
@@ -288,55 +316,115 @@ static bool isSerializable(id argument)
 
-    if (value.isArray) {
-        RetainPtr nsArray = [value toArray];
-        if (!nsArray)
+    if (JSValueIsDate(context, object))
+        return Seconds(JSValueToNumber(context, object, 0) / 1000.0);
+
+    if (JSValueIsArray(context, object)) {
+        SUPPRESS_UNCOUNTED_ARG JSValueRef lengthPropertyName = JSValueMakeString(context, adopt(JSStringCreateWithUTF8CString("length")).get());
+        JSValueRef lengthValue = JSObjectGetPropertyForKey(context, object, lengthPropertyName, nullptr);
+        double lengthDouble = JSValueToNumber(context, lengthValue, nullptr);
+        if (lengthDouble < 0 || lengthDouble > static_cast<double>(std::numeric_limits<size_t>::max()))
+            return NullType::NullPointer;
+
+        size_t length = lengthDouble;
+        Vector<JSObjectID> vector;
+        if (!vector.tryReserveInitialCapacity(length))
             return NullType::NullPointer;
 
-        Vector<JSObjectID> vector([nsArray count], [&](size_t i) {
-            return addObjectToMap([nsArray objectAtIndex:i]);
-        });
+        for (size_t i = 0; i < length; ++i)
+            vector.append(addObjectToMap(context, JSObjectGetPropertyAtIndex(context, object, i, nullptr)));
         return WTFMove(vector);
     }
 
-    RetainPtr object = checked_objc_cast<NSDictionary>([value toObject]);
+    JSPropertyNameArrayRef names = JSObjectCopyPropertyNames(context, object);
+    size_t length = JSPropertyNameArrayGetCount(names);
     HashMap<JSObjectID, JSObjectID> map;
-    for (id key in object.get())
-        map.add(addObjectToMap(key), addObjectToMap([object objectForKey:key]));
+    for (size_t i = 0; i < length; i++) {
+        JSRetainPtr<JSStringRef> key = JSPropertyNameArrayGetNameAtIndex(names, i);
+        SUPPRESS_UNCOUNTED_ARG map.add(addObjectToMap(context, JSValueMakeString(context, key.get())), addObjectToMap(context, JSObjectGetPropertyForKey(context, object, JSValueMakeString(context, key.get()), nullptr)));
+    }
+    JSPropertyNameArrayRelease(names);
     return WTFMove(map);
 }
 
 JavaScriptEvaluationResult::JavaScriptEvaluationResult(JSGlobalContextRef context, JSValueRef value)
-    : m_root(addObjectToMap([JSValue valueWithJSValueRef:value inContext:[JSContext contextWithJSGlobalContextRef:context]]))
+    : m_root(addObjectToMap(context, value))
 {
     m_jsObjectsInMap.clear();
     m_nullObjectID = std::nullopt;
 }
 
+JSValueRef JavaScriptEvaluationResult::toJS(JSGlobalContextRef context, Variant&& root)
+{
+    return WTF::switchOn(WTFMove(root), [&] (NullType) -> JSValueRef {
+        return JSValueMakeNull(context);
+    }, [&] (bool value) -> JSValueRef {
+        return JSValueMakeBoolean(context, value);
+    }, [&] (double value) -> JSValueRef {
+        return JSValueMakeNumber(context, value);
+    }, [&] (String&& value) -> JSValueRef {
+        auto string = OpaqueJSString::tryCreate(WTFMove(value));
+        return JSValueMakeString(context, string.get());
```

Comment:
Why not just use `JSStringCreateWithUTF8CString` instead?

Reply:
This avoids an unnecessary Latin1/UTF-16->UTF-8->Latin1/UTF-16 conversion roundtrip.